dsp/transform/hilbert.go

```diff
@@ -0,0 +1,89 @@
+// Copyright ©2024 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package transform
+
+import (
+	"math/cmplx"
+
+	"gonum.org/v1/gonum/dsp/fourier"
+)
+
+// Implements an approximate Hilbert transform that allows the user to determine an
+// approximate analytical signal of a real signal, and determine the real 'envelope'
+// of a signal.
+//
+// The underlying implementation uses a DFT and IDFT transform, so usual methods for
+// speeding up DFT/IDFT are likely to apply here as well.
+type Hilbert struct {
+	fft  *fourier.CmplxFFT
+	work []complex128
+}
+
+// Return a new Hilbert transformer for signals of size N.
+func NewHilbert(n int) *Hilbert {
+	return &Hilbert{
+		fft:  fourier.NewCmplxFFT(n),
+		work: make([]complex128, n),
+	}
+}
+
+// Retrieve the length N of signals that are valid input for this Hilbert transform.
+func (h *Hilbert) Len() int {
+	return len(h.work)
+}
+
+// Compute the analytical signal of a real signal, and stores the result in the dst slice,
+// returning it.
+// If the dst slice is nil, a new slice will be created and returned. The dst slice
+// must be the same length as the input signal.
+func (h *Hilbert) AnalyticSignal(signal []float64, dst []complex128) []complex128 {
+	if dst == nil {
+		dst = make([]complex128, len(signal))
+	}
+
+	for i, v := range signal {
+		h.work[i] = complex(v, 0)
+	}
+
+	// forward fft of signal
+	coeff := h.fft.Coefficients(dst, h.work)
+	for i := range h.work {
+		h.work[i] = 0
+	}
+
+	// multiply positive frequencies by 2, zero out negative frequencies
+	// however, leave dc unchanged (and nyquist when n%2 == 0)
+	h.work[0] = coeff[0]
+	for i, d := range coeff[1 : len(coeff)/2+1] {
+		h.work[i+1] = d * 2
+	}
+	if len(coeff)%2 == 0 {
+		h.work[len(coeff)/2] = coeff[len(coeff)/2]
+	}
+
+	// normalize the results for now
+	// TODO: is this important for analytic signal or envelope?
+	unnorm := h.fft.Sequence(dst, h.work)
+	for i, u := range unnorm {
+		unnorm[i] = u / complex(float64(len(unnorm)), 0)
+	}
+	return unnorm
+}
+
+// Compute the positive envelope of a real signal, and stores the result in the dst slice,
+// returning it.
+// If the dst slice is nil, a new slice will be created and returned. The dst slice
+// must be the same length as the input signal.
+func (h *Hilbert) Envelope(signal []float64, dst []float64) []float64 {
+	if dst == nil {
+		dst = make([]float64, len(signal))
+	}
+
+	analytic := h.AnalyticSignal(signal, nil)
+	for i, a := range analytic {
+		dst[i] = cmplx.Abs(a)
+	}
+	return dst
+}
```

Comment:
I'm not convinced this is needed. It's just a range loop.

Reply:
My only rebuttal is that some users would only be interested in the signal envelope, and possibly have only a fuzzy idea how it relates to the analytic signal (or even that it does, as I did when I first went looking for this). They would then have to look it up in order to recreate this loop, hoping that `AnalyticSignal` leaves the results in a state that all one needs to do is `Abs` each value. (For example, Gonum's FFT does not normalize its results, which is documented but surprising). With this method, it's at least searchable that Gonum can easily be used to compute a signal envelope regardless of how much expertise one has with DSP.

As an alternative, though, perhaps I could add to the doc comment of `AnalyticSignal`: "The envelope of the signal can be straightforwardly computed from the results using `cmplx.Abs` on each element".

Reply:
Suggest then having this in an example test. This shows the user how to use what we have and provides the code if they want to just copy/paste.

Reply:
Further to this, I think a more integrated example test would be worthwhile. This would be one that works through a toy example showing how the whole thing would be used.

Also, I have just noticed that this is a new package. It should have a doc.go file with package godoc explaining what the package is intended to cover.